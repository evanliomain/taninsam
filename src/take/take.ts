/**
 * @module array=>array
 */
import { isArray } from '../isArray';

/**
 * Extract n first elements of an array
 * @param n The number of elements to extract first
 * @return the function to apply on the array to extract its n first element
 * @example
 * ```
 * take<number>(2)([1, 2, 3, 4, 5]) // [1, 2]
 * take(2)('abba') // 'ab'
 * ```
 * @example Using the chain
 * ```
 * chain([1, 2, 3, 4, 5])
 *   .chain(take<number>(2))
 *   .value() // [1, 2]
 * ```
 */
export function take<T>(
  n: number
): (list: string | ReadonlyArray<T>) => string | ReadonlyArray<T> {
  return (list: string | ReadonlyArray<T>) => {
    if (isArray(list)) {
      return list.filter((_, index) => index < n);
    }
    return list.substring(0, n);
  };
}
