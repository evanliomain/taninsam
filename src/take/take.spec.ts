import { describe, expect, test } from 'vitest';
import { take } from './take';

describe('take function', () => {
  const inputArray = [1, 2, 3, 4, 5];
  const inputString = 'abba';

  test('is a pure function with array', () => {
    take(2)(inputArray);
    expect(inputArray).toMatchSnapshot();
  });

  test('is a pure function with string', () => {
    take(2)(inputString);
    expect(inputString).toMatchSnapshot();
  });

  test('[1, 2, 3, 4, 5] |> take(2) === [1, 2]', () => {
    expect(take(2)(inputArray)).toMatchSnapshot();
  });

  test('[1, 2] |> take(3) === [1, 2]', () => {
    expect(take(3)([1, 2])).toMatchSnapshot();
  });

  test('[1, 2] |> take(0) === []', () => {
    expect(take(0)([1, 2])).toMatchSnapshot();
  });

  test('"abba" |> take(2) === "ab"', () => {
    expect(take(2)(inputString)).toMatchSnapshot();
  });

  test('"ab" |> take(3) === "ab"', () => {
    expect(take(3)('ab')).toMatchSnapshot();
  });

  test('"abba" |> take(0) === ""', () => {
    expect(take(0)(inputString)).toMatchSnapshot();
  });
});
