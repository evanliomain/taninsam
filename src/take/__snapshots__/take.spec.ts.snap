// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`take function > "ab" |> take(3) === "ab" 1`] = `"ab"`;

exports[`take function > "abba" |> take(0) === "" 1`] = `""`;

exports[`take function > "abba" |> take(2) === "ab" 1`] = `"ab"`;

exports[`take function > [1, 2, 3, 4, 5] |> take(2) === [1, 2] 1`] = `
[
  1,
  2,
]
`;

exports[`take function > [1, 2] |> take(0) === [] 1`] = `[]`;

exports[`take function > [1, 2] |> take(3) === [1, 2] 1`] = `
[
  1,
  2,
]
`;

exports[`take function > is a pure function with array 1`] = `
[
  1,
  2,
  3,
  4,
  5,
]
`;

exports[`take function > is a pure function with string 1`] = `"abba"`;
